Asia Roadshow - Plushie NFT Collection

cybervoid0 opened the pull request:
# Asia Roadshow - Plushie NFT Collection

## Tickets
[Asia Roadshow - Plushie NFT Collection](https://rsklabs.atlassian.net/browse/DAO-1503)
[DAO-1516 Build new NFT contract](https://rsklabs.atlassian.net/browse/DAO-1516)

## What

This pull request introduces a new non-transferrable NFT smart contract for the Plushie Series 1 Rootstock Collective and updates the development environment to support its deployment and testing. The changes include the addition of the contract, configuration updates for Solidity compilation, improved VSCode integration for linting and formatting, and a new Ignition deployment module.

### Smart Contract Addition

* Added `PlushieSeries1RootstockCollective.sol`, an upgradeable, non-transferrable ERC721 contract for POAP-style NFTs with whitelisting, stRIF balance checks, and administrative controls.

### Development Environment & Tooling

* Updated `hardhat.config.ts` to support Solidity v0.8.30 with optimizer settings and expanded the number of test accounts for the Hardhat network. [[1]](diffhunk://#diff-57cf5378f8cab20b02b279097ba6fad08eb53f747d81770045dda9c5f95effe3R21-R26) [[2]](diffhunk://#diff-57cf5378f8cab20b02b279097ba6fad08eb53f747d81770045dda9c5f95effe3R44-R46)
* Added a Hardhat Ignition deployment module (`PlushieSeries1Module.ts`) to automate contract and proxy deployment with parameterized initialization.

### Editor & Linting Improvements

* Enhanced `.vscode/settings.json` to enforce ESLint fixes and Prettier formatting on save for TypeScript and JavaScript files, and validated ESLint for these languages.

cybervoid0 committed:
feat: add PlushieNFT contract and related modules

- Introduced `PlushieNFT` contract with role-based access control for minting.
- Added `ERC721NonTransferrable` base contract to disable token transfers.
- Created `PlushieNftModule` for deployment configuration.
- Implemented tests for role management and NFT functionality.
- Updated Hardhat configuration to include Solidity 0.8.30 and optimizer settings.
- Enhanced VSCode settings for TypeScript and JavaScript linting and formatting.

cybervoid0 committed:
refactor: update PlushieNFT contract and provide more tests

- Removed `ERC721NonTransferrable` base contract to simplify transfer logic.
- Updated `PlushieNFT` to include custom error handling for admin role violations and transfer prevention.
- Adjusted test cases to reflect changes in role management and minting logic.
- Modified Hardhat configuration to allow for 50 accounts in the local network.

cybervoid0 committed:
chore: rename contract and related modules to PlushieSeries1RootstockCollective, add StRif threshold check

cybervoid0 committed:
refactor: enhance PlushieSeries1RootstockCollective contract and tests

- Updated event parameters to use IERC20 for token references.
- Added custom error for invalid token address.
- Improved role management initialization and hierarchy.
- Modified base URI function to include folder IPFS CID.
- Implemented transfer restrictions in the NFT contract.
- Updated tests to reflect changes in token URI and transfer restrictions.

cybervoid0 committed:
refactor: simplify metadata handling in PlushieSeries1RootstockCollective contract

- Removed ERC721URIStorage to streamline token URI management.
- Updated tokenURI function to return a single metadata IPFS CID for all tokens.
- Adjusted tests to verify the new token URI structure and ensure consistency across minted tokens.

cybervoid0 committed:
doc: update comment in PlushieSeries1RootstockCollective contract

TravellerOnTheRun commented on contracts/NFT/PlushieSeries1RootstockCollective.sol:
```diff
@@ -0,0 +1,286 @@
+  // STATE VARIABLES. Storage packing optimization (4 slots total)
+
+  // Slot 1: 32 bytes
+  uint256 private _totalMinted;
```
Isn't 32 bytes too large for the purpose of this NFT? I know it's optimization is not necessary at all and won't save much, but realistically is it even reachable?

cybervoid0 replied:
OZ and most of the tools and collections expect uint256. Since token ID is uint256 by standard, there is no sense of making this a different type

TravellerOnTheRun commented on contracts/NFT/PlushieSeries1RootstockCollective.sol:
```diff
@@ -0,0 +1,286 @@
+
+  // Slot 3: 32 bytes
+  /// @notice Maximum number of tokens
+  uint256 public maxSupply;
```
Same goes here

cybervoid0 replied:
OZ and most of the tools and collections expect uint256. Since token ID is uint256 by standard, there is no sense of making this a different type

TravellerOnTheRun commented on contracts/NFT/PlushieSeries1RootstockCollective.sol:
```diff
@@ -0,0 +1,286 @@
+   * @dev The Minter role is revoked after minting to monitor the number of minters.
+   */
+  function mint() external onlyRole(MINTER_ROLE) returns (uint256) {
+    address caller = _msgSender();
```
"claimer" maybe?

cybervoid0 replied:
Don't nitpick please. A person who calls the function can be called caller and this name is clear enough

TravellerOnTheRun replied:
That's why it's a maybe

TravellerOnTheRun commented on contracts/NFT/PlushieSeries1RootstockCollective.sol:
```diff
@@ -0,0 +1,286 @@
+    address account
+  ) public virtual override(AccessControlUpgradeable, IAccessControl) {
+    if (role == DEFAULT_ADMIN_ROLE && getRoleMemberCount(DEFAULT_ADMIN_ROLE) >= 1) {
+      revert PlushieNftAdminRoleViolation();
```
Don't functions grantRole and renounce role contradict each other. Cause if admin cannot renounce the role if there's only one admin and admin cannot grantRole if there's already a one admin, does it mean that changing admin is impossible in that case?

cybervoid0 replied:
I added another function `transferDefaultAdminRole` specially for the case of transferring ownership. But I kept those 2 overrides because they guarantee that the Default admin is always only one

cybervoid0 replied:
and I added a test case showing that the transferDefaultAdminRole function really does what it supposed to

TravellerOnTheRun commented on test/PlushieSeries1.test.ts:
```diff
@@ -0,0 +1,565 @@
+
+  describe('upon deployment', () => {
+    it('should set up proper NFT name and symbol', async () => {
+      expect(await plushie.name()).to.equal('PlushieSeries1RootstockCollective')
```
Save name to const?

cybervoid0 replied:
it is used only once! don't nitpick

TravellerOnTheRun commented on test/PlushieSeries1.test.ts:
```diff
@@ -0,0 +1,565 @@
+  describe('upon deployment', () => {
+    it('should set up proper NFT name and symbol', async () => {
+      expect(await plushie.name()).to.equal('PlushieSeries1RootstockCollective')
+      expect(await plushie.symbol()).to.equal('PS1')
```
Symbol to const?

cybervoid0 replied:
it is used only once! don't nitpick

TravellerOnTheRun commented on test/PlushieSeries1.test.ts:
Add test to change the admins from one to another

cybervoid0 replied:
yes

cybervoid0 committed:
feat: add transferDefaultAdminRole function and update error handling in PlushieSeries1RootstockCollective contract

- Introduced transferDefaultAdminRole function to allow the admin to transfer their role to another address.
- Updated error handling to use a more generic PlushieNftInvalidAddress for invalid address checks in the contract.
- Enhanced tests to verify the functionality of the new role transfer feature and its impact on admin privileges.

cybervoid0 committed:
feat: add documentation for transferDefaultAdminRole function in PlushieSeries1RootstockCollective contract

- Added detailed comments to describe the purpose and functionality of the transferDefaultAdminRole function.
- Clarified the atomic nature of the role transfer process and the parameters involved.

TravellerOnTheRun reviewed (approved):
Looks good, let's transfer to security
